Add support for Azure Event Hubs

oising opened the pull request:
Initial support for Azure Event Hubs, as per https://github.com/dotnet/aspire/issues/1147

- [x] Minimally adapt the Service Bus component and resource for Event Hubs
- [x] Add top-level support for EventHubProducerClient 
- [x] Add top-level support for EventHubConsumerClient
- [x] Add top-level support for EventProcessorClient
  - [x] Allow providing component name for a BlobContainerClient to use (read connection string and instantiate local instance)
  - [ ] Read BlobContainerClient from keyed DI using its Aspire connection name
- [x] Add top-level support for PartitionReceiver
- [x] Add playground sample AppHost and projects (modify davidfowl's test project)
- [x] Add ability to configure credential used in Azure Provisioner (via AzureProvisionCredential in AppHost configuration)
- [x] Add xmldoc documentation for components etc
- [x] Update README

Optional extras (may defer to a second PR)

- [x] Add support for chained AddHub(...) calls for top-level EH resource to allow variable config of hubs
  - [ ] Implement as child resource to allow deep-linking (i.e. provide connection string including EntityPath) 
- [ ] Implement Microsoft.Extensions.Azure style top-level component AddEventHubs(...) and hang clients configuration from clientBuilder
- [ ] Add Streaming sample with Orleans

## Aspire API shortcomings

EventProcessorClient requires a BlobContainerClient for managing EH checkpoints. Ideally we should be able to pass a connection name to a mounted aspire blob client component and retrieve the instance from the service provider, but there is no plumbing to get it in the component infra. The unbuilt service collection is available on the distributed builder and it could be fished out of there, assuming it was added before the current component, but this seems flakey. Instead, I'm creating a new blob client using the blob connection name (taken from user-supplied settings) and creating my own. 

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2870)

davidfowl commented:
The azure provisioning stuff here is outdated so I would just delete all of it since it isn't required.

davidfowl commented:
This entire PR seems based on a really old version of aspire. Everything is bicep based now.

oising commented:
> This entire PR seems based on a really old version of aspire. Everything is bicep based now.

I think your definition of "really old" doesn't match with mine :) I'm working off of main, and I cloned barely a week ago? I'll update it.

davidfowl commented:
I'm not sure what happened, but if you look at main the Aspire.Hosting.Azure.Provisioning package has been deleted... 😄

oising commented:
Restarting for the third time, because of https://github.com/dotnet/aspire/pull/2875

Sigh.

oising committed:
initial commit for event hubs component and resource (non-functional, but skeleton is there)

oising committed:
corrected roledefinition

oising committed:
added consumer client (now have producer, consumer and processor components); modified davidfowl's demo EH project to use new resources and components; bugfixes;

davidfowl commented on playground/AspireEventHub/EventHubs.ServiceDefaults/Extensions.cs:
```diff
@@ -0,0 +1,116 @@
+using Microsoft.AspNetCore.Builder;
```
Remove this and use [the shared service defaults](https://github.com/dotnet/aspire/tree/2788dd5a66ae4902f9d1a6a08c19f71359b7ea95/playground/Playground.ServiceDefaults)

oising replied:
ah, I think I'm already referencing the shared playground project but I didn't notice this. thanks.

oising committed:
remove obsolete servicedefaults project

oising committed:
Merge branch 'main' into support-eventhubs

davidfowl commented:
@sebastienros @eerhardt for the component review  
@mitchdenny @JoshLove-msft for the resource review. CDK codegen is also required here.

davidfowl commented:
Also @jsquire for azure sdk review

oising committed:
fixes for processor client; updated build props/targets for playground project

oising commented:
I rebased to stay up to date and now I'm hitting:

Application orchestrator dependency check had an unexpected error System.TimeoutException: The operation has timed out.
   at Aspire.Hosting.Dcp.DcpDependencyCheck.EnsureDcpDependenciesAsync(CancellationToken cancellationToken) in C:\p\github\aspire\src\Aspire.Hosting\Dcp\DcpDependencyCheck.cs:line 67

on start up.

davidfowl commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -23,21 +26,44 @@ protected override void BindClientOptionsToConfiguration(IAzureClientBuilder<Eve
-        }, requiresCredential: false);
+                // todo: add more settings and clientoptions validation; also we should use a deterministic ID for the processor
+
+                var blobConnectionString =
+                    builderConfiguration.GetConnectionString(settings.BlobClientConnectionName);
+
+                Uri blobUri = new Uri($"{blobConnectionString}/checkpoints"); // fixme: this should be a compound name including the consumer group name ideally
+                var blobClient = new BlobContainerClient(blobUri, cred);
+                blobClient.CreateIfNotExists(); // this works, but the processor client fails to create blobs later?
```
Instead, this logic should inject the keyed blob client

oising replied:
Yeah, I agree - but the plumbing is not there. I added comments above in the main body of the PR (aspire api shortcomings)

oising replied:
> Instead, this logic should inject the keyed blob client or non keyed (🤔)

Good luck trying to untangle the IAzureClient* APIs to inject anything supplementary -- it's a jungle of extension methods; otherwise I would have suggested something. It seems to me that some deep work may be needed to allow inter-component dependencies.

davidfowl commented:
> Application orchestrator dependency check had an unexpected error System.TimeoutException: The operation has timed out.
at Aspire.Hosting.Dcp.DcpDependencyCheck.EnsureDcpDependenciesAsync(CancellationToken cancellationToken) in C:\p\github\aspire\src\Aspire.Hosting\Dcp\DcpDependencyCheck.cs:line 67

Consistently?

oising commented:
> > Application orchestrator dependency check had an unexpected error System.TimeoutException: The operation has timed out.
> > at Aspire.Hosting.Dcp.DcpDependencyCheck.EnsureDcpDependenciesAsync(CancellationToken cancellationToken) in C:\p\github\aspire\src\Aspire.Hosting\Dcp\DcpDependencyCheck.cs:line 67
> 
> Consistently?

Yes. I tried bumping the timeout to 5 mins and at some point something else timed out and brought the whole process down. I tried disabling the timeout (setting to 0) and same thing. Tried about a half dozen times.

mitchdenny commented on playground/AspireEventHub/EventHubsApi/Program.cs:
```diff
@@ -0,0 +1,26 @@
+
+builder.AddServiceDefaults();
+
+builder.AddAzureEventHubProducerClient("eventhubns", settings =>
```
One of the things that I think makes the whole experience with Aspire better is when we can "deep link" to the resource that we want to use. In the settings callback here you are configuring the name of the hub that you want to set up the producer to communicate with, but I'd want to see the ability to also inject an `EventHubProducerClient` based on purely the data that comes through in a connection string.

mitchdenny committed:
Update AzureEventHubsExtensions.cs

No need to construct this (we recently changed this in the others).

mitchdenny commented on src/Aspire.Hosting.Azure/AzureEventHubsResource.cs:
```diff
@@ -0,0 +1,29 @@
+    /// Gets the connection string template for the manifest for the Azure Event Hubs endpoint.
+    /// </summary>
+    public ReferenceExpression ConnectionStringExpression =>
+        ReferenceExpression.Create($"{ServiceBusEndpoint}");
```
Why not use the `Endpoint=sb://...` variant? That way could append the entity path?

oising replied:
I could -- but I'm duplicating what the Service Bus resource/component does. If we change this, we should change the ASB interfaces too. As you know, EH and ASB have the same lineage. It might be better to plan enhancements like this as a batch and update ASB and EH in the same PR. What do you think?

mitchdenny commented on src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:
```diff
@@ -256,13 +256,14 @@ private async Task ProcessResourceAsync(IConfiguration configuration, Lazy<Task<
+        //     ExcludeAzurePowerShellCredential = true,
+        //     CredentialProcessTimeout = TimeSpan.FromSeconds(15)
+        // });
+        var credential = new AzureCliCredential();
```
???

oising replied:
my personal azure subscription where I test things is not related to the ambient login/visualstudio login I have on my work laptop, which is linked to a different tenant. Using the DefaultAzureCredential means it hits my work subscription first. I spoke to @davidfowl about this -- there needs to be a way to use a credential other than the default for deployment. This is a temporary hack :)

oising replied:
![image](https://github.com/dotnet/aspire/assets/1844001/ab4f3eb3-d522-47ae-b56d-c6f21278d598)

mitchdenny commented on src/Components/Aspire.Azure.Messaging.EventHubs/README.md:
````diff
@@ -0,0 +1,145 @@
+Then, in the _Program.cs_ file of `AppHost`, add a Service Bus connection and consume the connection using the following methods:
+
+```csharp
+var serviceBus = builder.AddAzureServiceBus("sb");
````
Needs to be updated for event hubs.

oising replied:
Done

mitchdenny commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <param name="name">The name of the hub.</param>
+    /// <param name="configureQueue">Optional callback to customize the hub.</param>
+    [RequiresPreviewFeatures]
+    public static IResourceBuilder<AzureEventHubsResource> AddHub(this IResourceBuilder<AzureEventHubsResource> builder,
```
Should AzureEventHubs have a child resource so that you could also setup consumer groups or do we expect people to configure alternative consumer groups via the CDK callback?

oising replied:
Might push back on this idea until we could implement the same for ASB queues and topics also.

mitchdenny commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <param name="builder">The Azure Event Hubs resource builder.</param>
+    /// <param name="name">The name of the hub.</param>
+    /// <param name="configureQueue">Optional callback to customize the hub.</param>
+    [RequiresPreviewFeatures]
```
`[Experimental]` (changed today)

mitchdenny replied:
We have some already in the code base for this exact scenario, just copy them (search for Experimental).

mitchdenny commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+        this IDistributedApplicationBuilder builder, string name)
+    {
+#pragma warning disable CA2252 // This API requires opting into preview features
+        return builder.AddAzureEventHubs(name, (_, _, _) => { });
```
Change the callback to null since the argument is nullable and you are using `?.Invoke(...)`.

mitchdenny commented:
This looks like a really good start. My main piece of feedback is around what level of developer experience we are going for. One of the things that I like about things like our database resource types is that I can do this in the app host:

```csharp
var db = builder.AddSqlServer("sql").AddDatabase("db");
```

... and this in the service project:

```csharp
builder.AddSqlClient("db");
```

... and then the service is available ready to use. With services such as Cosmos DB, and this resource we break that model because methods such as `AddHub` don't create a child resource that enables its connection string to deep link to the hub (e.g. EntityPath in the standard connection string).

This means that every time I want to use an event hub I have to provide a connection string to the namespace, but also the hub name and optionally the consumer group. I wish we could improve this experience because it steepens the learning curve.

jsquire commented on playground/AspireEventHub/EventHubsConsumer/Processor.cs:
```diff
@@ -0,0 +1,45 @@
+            return Task.CompletedTask;
+        };
+
+        await client.StartProcessingAsync(cancellationToken);
```
We should include comments that discuss the threading model for the processor and guidance for partition ownership.  This pattern is great for a simple sample but goes against how the processor is intended to be used for the majority of production scenarios.

Too often we see developers attempt to handle too many partitions on a single machine and see the expected poor performance as a result.

oising replied:
Good call. I'll link to the guidance and insert a pertinent quote as a comment.

oising replied:
Done.

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    {
+        var configureConstruct = (ResourceModuleConstruct construct) =>
+        {
+            var serviceBusNamespace = new EventHubsNamespace(construct, name: name);
```
nit:  The variable names should reflect Event Hubs, not Service Bus.

oising replied:
I go back and forth about this since all event hubs namespaces look like *.servicebus.windows.net -- even the Bicep templates expose it as a serviceBusNamespace, right?

oising replied:
but I'll change it to be consistent with the dotnet type.

jsquire replied:
While they share a common lineage, Event Hubs namespaces and Service Bus namespaces are two different Azure resources and should be treated  interpedently.   I can't speak to what the Aspire bicep files are doing, but it would be based on a `listKeys` call to the Event Hubs resource.

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// Adds an Azure Event Hubs hub resource to the application model. This resource requires an <see cref="AzureEventHubsResource"/> to be added to the application model.
+    /// </summary>
+    /// <param name="builder">The Azure Event Hubs resource builder.</param>
+    /// <param name="name">The name of the hub.</param>
```
```suggestion
    /// <param name="name">The name of the Event Hub.</param>
```

nit: Branding.   Leaving one comment, but it applies to all usages.

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <param name="name">The name of the hub.</param>
+    /// <param name="configureQueue">Optional callback to customize the hub.</param>
+    [RequiresPreviewFeatures]
+    public static IResourceBuilder<AzureEventHubsResource> AddHub(this IResourceBuilder<AzureEventHubsResource> builder,
```
```suggestion
    public static IResourceBuilder<AzureEventHubsResource> AddEventHub(this IResourceBuilder<AzureEventHubsResource> builder,
```

nit: Branding

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <param name="configureQueue">Optional callback to customize the hub.</param>
+    [RequiresPreviewFeatures]
+    public static IResourceBuilder<AzureEventHubsResource> AddHub(this IResourceBuilder<AzureEventHubsResource> builder,
+        string name, Action<IResourceBuilder<AzureEventHubsResource>, ResourceModuleConstruct, EventHub>? configureQueue)
```
```suggestion
        string name, Action<IResourceBuilder<AzureEventHubsResource>, ResourceModuleConstruct, EventHub>? configureEventHub)
```

There's no such concept of a queue in Event Hubs.

oising replied:
Good catch. Boilerplate copypasta error.

jsquire commented on src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs:
```diff
@@ -0,0 +1,129 @@
+                configureClientBuilder, connectionName: name, serviceKey: name);
+    }
+
+    public static void AddAzureEventHubConsumerClient(
```
We should include the PartitionReceiver as well.

oising replied:
Working on this now.

oising replied:
Done

jsquire commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -0,0 +1,69 @@
+
+                Uri blobUri = new Uri($"{blobConnectionString}/checkpoints"); // fixme: this should be a compound name including the consumer group name ideally
+                var blobClient = new BlobContainerClient(blobUri, cred);
+                blobClient.CreateIfNotExists(); // this works, but the processor client fails to create blobs later?
```
> blobClient.CreateIfNotExists(); // this works, but the processor client fails to create blobs later?

The processor intentionally does not create the container, so this would be needed if you're intending to create automatically.   Note that this does require permissions above the container to work.

oising replied:
@jsquire Yep, I understand that the processor makes no attempt to create the container. I am actually referring to a different problem I am experiencing. If the blob container client uses a tokenidentity (and not via connectionstring) to authenticate, the processor client fails to read/create blob checkpoints with "400 Value for one of the query parameters specified in the request URI is invalid." It's very suspicious.

oising replied:
^ Looking at all of the examples on learn.microsoft.com and on github, they all configure a processor with a sasuri or sharedkey connectionstring. I will try to repro independently, unless there's a known issue?

jsquire replied:
We test nightly with token credentials as well as with connection strings.   Can't say that I've ever seen or heard of this before.   If you're able to repro and can grab the exact error message, full stack trace, and capture Azure SDK logs for that, it would help me pull the Storage team in to troubleshoot.   [Logging sample](https://learn.microsoft.com/dotnet/azure/sdk/logging)

We also have samples for identity-based connections.   For example, [this one](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/samples/Sample05_IdentityAndSharedAccessCredentials.md#processing-events-with-identity-based-authorization).

oising replied:
![image](https://github.com/dotnet/aspire/assets/1844001/9215e47e-0803-4e4a-859c-6cce46d954ae)

This is as bare bones as it gets. Very odd.

oising replied:
I switched to using the BlobUriBuilder as an example and now it works. 

```csharp
                BlobUriBuilder blobUriBuilder = new BlobUriBuilder(new Uri(blobConnectionString!));
                blobUriBuilder.BlobContainerName = "checkpoints";
                Uri blobUri = blobUriBuilder.ToUri();

                //Uri blobUri = new Uri($"{blobConnectionString}/checkpoints"); 
```

The reason is that my interpolated blobUri had two slashes before checkpoints. I did not think that would parse a valid Uri, but apparently so! But that leads to the 400 error.

jsquire replied:
Any chance you can grab the message and stack trace after flattening the aggregate?

oising replied:
> Any chance you can grab the message and stack trace after flattening the aggregate?

```
System.AggregateException
  HResult=0x80131500
  Message=One or more errors occurred. (Service request failed.
Status: 400 (Value for one of the query parameters specified in the request URI is invalid.)
ErrorCode: InvalidQueryParameterValue

Headers:
Transfer-Encoding: chunked
Server: Microsoft-HTTPAPI/2.0
x-ms-request-id: ee637926-401e-003e-3571-796f9f000000
x-ms-client-request-id: a54c06b7-723e-4236-af03-4f8ce11677db
x-ms-error-code: InvalidQueryParameterValue
Date: Mon, 18 Mar 2024 20:15:56 GMT
)
  Source=Azure.Messaging.EventHubs
  StackTrace:
   at Azure.Messaging.EventHubs.Primitives.EventProcessor`1.<StartProcessingInternalAsync>d__86.MoveNext()
   at Azure.Messaging.EventHubs.Primitives.EventProcessor`1.<StartProcessingAsync>d__61.MoveNext()
   at Azure.Messaging.EventHubs.EventProcessorClient.<StartProcessingInternalAsync>d__73.MoveNext()
   at Azure.Messaging.EventHubs.EventProcessorClient.<StartProcessingAsync>d__54.MoveNext()
   at EventHubsConsumer.Processor.<StartAsync>d__3.MoveNext() in C:\p\github\aspire\playground\AspireEventHub\EventHubsConsumer\Processor.cs:line 31
   at Microsoft.Extensions.Hosting.Internal.Host.<<StartAsync>b__15_1>d.MoveNext()
   at Microsoft.Extensions.Hosting.Internal.Host.<ForeachService>d__18`1.MoveNext()
   at Microsoft.Extensions.Hosting.Internal.Host.<StartAsync>d__15.MoveNext()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.<RunAsync>d__4.MoveNext()
   at Program.<<Main>$>d__0.MoveNext() in C:\p\github\aspire\playground\AspireEventHub\EventHubsConsumer\Program.cs:line 42

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
RequestFailedException: Service request failed.
Status: 400 (Value for one of the query parameters specified in the request URI is invalid.)
ErrorCode: InvalidQueryParameterValue

Headers:
Transfer-Encoding: chunked
Server: Microsoft-HTTPAPI/2.0
x-ms-request-id: ee637926-401e-003e-3571-796f9f000000
x-ms-client-request-id: a54c06b7-723e-4236-af03-4f8ce11677db
x-ms-error-code: InvalidQueryParameterValue
Date: Mon, 18 Mar 2024 20:15:56 GMT
```

oising replied:
I guess this is a bug because the blob container client call  `blobClient.CreateIfNotExists()` succeeds and creates a `checkpoints` container in the root despite the double slashes, but the processor chokes.

oising replied:
Ooops, the inner stacktrace is:

```
   at Azure.Core.HttpPipelineExtensions.<ProcessMessageAsync>d__0.MoveNext()
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at Azure.Storage.Blobs.BlobRestClient.<GetPropertiesAsync>d__14.MoveNext()
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<GetPropertiesInternal>d__124.MoveNext()
   at Azure.Storage.Blobs.Specialized.BlobBaseClient.<GetPropertiesAsync>d__123.MoveNext()
   at Azure.Messaging.EventHubs.Primitives.BlobCheckpointStoreInternal.<GetCheckpointAsync>d__20.MoveNext()
   at Azure.Messaging.EventHubs.EventProcessorClient.<GetCheckpointAsync>d__66.MoveNext()
   at Azure.Messaging.EventHubs.Primitives.EventProcessor`1.<ValidateStorageConnectionAsync>d__95.MoveNext()
   at Azure.Messaging.EventHubs.Primitives.EventProcessor`1.<ValidateProcessingPreconditions>d__72.MoveNext()
   ```

jsquire commented on src/Components/Aspire.Azure.Messaging.EventHubs/README.md:
```diff
@@ -0,0 +1,145 @@
+# Aspire.Azure.Messaging.ServiceBus
+
+Registers a [ServiceBusClient](https://learn.microsoft.com/dotnet/api/azure.messaging.servicebus.servicebusclient) in the DI container for connecting to Azure Service Bus.
```
Looks like this is all still Service Bus content, so skipping for now.

oising replied:
Yep, copypasta from the ASB bits. I'll update soon.

oising commented on playground/AspireEventHub/EventHubsConsumer/Consumer.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        void LogString(string message) => logger.Log(LogLevel.Information, 0, message, null, (s, _) => s);
+
+        await foreach (var partition in client.ReadEventsAsync(stoppingToken))
```
Good idea. Will do. It's just here for completeness's sake.

oising commented on src/Aspire.Hosting.Azure/AzureEventHubsResource.cs:
```diff
@@ -0,0 +1,29 @@
+    internal List<(string Name, Action<IResourceBuilder<AzureEventHubsResource>, ResourceModuleConstruct, EventHub>? Configure)> Hubs { get; } = [];
+
+    /// <summary>
+    /// Gets the "serviceBusEndpoint" output reference from the bicep template for the Azure Storage resource.
```
I copied a lot of the boilerplate from the ASB bits. I'll correct it there too.

Commit:
Merge branch 'main' into support-eventhubs

oising committed:
some minor refactoring of processor client; made playground project more robust

oising committed:
Merge branch 'main' into support-eventhubs

oising committed:
address review items; automatically generate processor identifier from hub and consumergroup; add more comments; minor refactoring

oising committed:
merge upstream main

oising commented:
> I rebased to stay up to date and now I'm hitting:
> 
> Application orchestrator dependency check had an unexpected error System.TimeoutException: The operation has timed out. at Aspire.Hosting.Dcp.DcpDependencyCheck.EnsureDcpDependenciesAsync(CancellationToken cancellationToken) in C:\p\github\aspire\src\Aspire.Hosting\Dcp\DcpDependencyCheck.cs:line 67
> 
> on start up.

See https://github.com/dotnet/aspire/issues/3003

Solved, but we need something in release notes for this issue.

oising committed:
fix solution merge dumbness

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+
+            eventHubsNamespace.AssignProperty(p => p.Sku.Name, new Parameter("sku", defaultValue: "Standard"));
+
+            var serviceBusDataOwnerRole = eventHubsNamespace.AssignRole(RoleDefinition.EventHubsDataOwner);
```
```suggestion
            var eventHubsDataOwnerRole = eventHubsNamespace.AssignRole(RoleDefinition.EventHubsDataOwner);
```

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+            var serviceBusDataOwnerRole = eventHubsNamespace.AssignRole(RoleDefinition.EventHubsDataOwner);
+            serviceBusDataOwnerRole.AssignProperty(p => p.PrincipalType, construct.PrincipalTypeParameter);
+
+            eventHubsNamespace.AddOutput("serviceBusEndpoint", sa => sa.ServiceBusEndpoint);
```
```suggestion
            eventHubsNamespace.AddOutput("eventHubsEndpoint", sa => sa.ServiceBusEndpoint);
```

jsquire commented on src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+            var serviceBusDataOwnerRole = eventHubsNamespace.AssignRole(RoleDefinition.EventHubsDataOwner);
+            serviceBusDataOwnerRole.AssignProperty(p => p.PrincipalType, construct.PrincipalTypeParameter);
+
+            eventHubsNamespace.AddOutput("serviceBusEndpoint", sa => sa.ServiceBusEndpoint);
```
@JoshLove-msft:   Is the CDK sending back "ServiceBusEndpoint" for an Event Hubs resource?   This looks wrong.

davidfowl replied:
That's the name of the property on the ARM API AFAIK

oising replied:
I'm not the only one guilty of this :) 

![image](https://github.com/dotnet/aspire/assets/1844001/948c54d5-edfa-4dc3-b1f9-8205f25f4524)

oising replied:
I've updated all the Bicep references and any of my code to use eventHubsEndpoint instead of serviceBusEndpoint, even though that feels weird to me, perversely. But we have to start somewhere :)

jsquire replied:
Looks like you're right.  _([example](https://github.com/Azure/azure-rest-api-specs/blob/969e0846e56a0869203bcc52773415c71115f59e/specification/eventhub/resource-manager/Microsoft.EventHub/stable/2021-11-01/examples/NameSpaces/EHNameSpaceCreate.json#L86))_   
That's an odd choice, but I agree we should keep it aligned with ARM.

oising replied:
So @davidfowl -- do I revert back to ServiceBusEndpoint/serviceBusEndpoit for outputs, vars etc? You seem to indicate the opposite with the thumbs up on my change below :)  -- @jsquire you agree with who? I am confused lol

jsquire replied:
@oising:  Sorry, our comments cross-threaded, David's comment was the last visible one and the one that I was responding to.   I defer to you and David on the path forward, but for my two cents, I like your updates.  😄

jsquire commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -0,0 +1,95 @@
+                }
+
+                // configure processor identifier
+                options.Identifier ??= $"{settings.EventHubName}-{settings.ConsumerGroup ?? "default"}";
```
This is a nice touch, though we'll want something that can be unique otherwise we'll see problems with load balancing and ownership.   Maybe stick the machine name and a random slug in here too?   

That'll help keep a meaningful name for logs, but also ensure that we don't see collisions for processors running in the same cluster.

oising replied:
This look okay to you?

     var slug = Guid.NewGuid().ToString().Substring(24);
     options.Identifier ??= $"{Environment.MachineName}-{settings.EventHubName}-{settings.ConsumerGroup ?? "default"}-{slug}";

jsquire replied:
Yup, that looks good!

oising committed:
Update src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs

Co-authored-by: Jesse Squire <jesse.squire@gmail.com>

oising committed:
Update src/Aspire.Hosting.Azure/Extensions/AzureEventHubsExtensions.cs

Co-authored-by: Jesse Squire <jesse.squire@gmail.com>

oising committed:
update EH endpoint references to use eventHubsEndpoint instead of serviceBusEndpoint; beef up processor identifier naming.

davidfowl commented on src/Aspire.Hosting.Azure/AzureEventHubsResource.cs:
```diff
@@ -17,13 +17,13 @@ public class AzureEventHubsResource(string name, Action<ResourceModuleConstruct>
+    /// Gets the "eventHubsEndpoint" output reference from the bicep template for the Azure Event Hubs resource.
     /// </summary>
-    public BicepOutputReference ServiceBusEndpoint => new("serviceBusEndpoint", this);
+    public BicepOutputReference EventHubsEndpoint => new("eventHubsEndpoint", this);
```
👍🏾

oising replied:
I'm going to break the mold here and keep it as eventHubsEndpoint unless otherwise advised. The experienced people will realize under the covers it's actually serviceBusEndpoint, and the newbies will just be less confused (until they find out the former lol)

oising commented:
I've cleaned up my credential hack in the Azure provisioner to allow those of us with more complex Azure environments to choose a specific credential for the provisioner. We can change it later if someone comes up with something better, but with this we can configure it in AppHost appsettings.

![image](https://github.com/dotnet/aspire/assets/1844001/44d3e422-1f24-460a-80a0-ffc3eadca673)

oising committed:
add ability to configure credential for azure provisioner (removed my hack); added first batch of documentation for component; added partitionreceiver component.

davidfowl commented:
Can we undraft this PR now @oising ?

davidfowl commented on src/Aspire.Hosting.Azure/Provisioning/Provisioners/AzureProvisioner.cs:
```diff
@@ -266,13 +266,25 @@ private async Task ProcessResourceAsync(IConfiguration configuration, Lazy<Task<
 
     private async Task<ProvisioningContext> GetProvisioningContextAsync(string? userSecretsPath, CancellationToken cancellationToken)
```
@oising this fixes https://github.com/dotnet/aspire/issues/1329, which is currently assigned to me. Do you want me to take this change separately?

oising replied:
Would you prefer I create a different PR for that? I don't mind doing the work.

davidfowl replied:
Yes please can you split it out?

oising replied:
@davidfowl  https://github.com/dotnet/aspire/pull/3083

oising commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -42,8 +45,21 @@ public sealed class AzureMessagingEventHubsSettings : IConnectionStringSettings
+    /// Gets or sets the event position to start from in the bound partition. Defaults to <see cref="EventPosition.Latest" />.
+    /// </summary>
+    /// <remarks>Applies only to <see cref="PartitionReceiver"/></remarks>
+    public EventPosition EventPosition { get; set; } = EventPosition.Latest;
```
Ah, I misremembered. I could have swore it was latest as default elsewhere. Ooops.

oising committed:
Update src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs

Co-authored-by: Jesse Squire <jesse.squire@gmail.com>

oising committed:
Merge branch 'main' into support-eventhubs

davidfowl commented:
The grand split has happened!

oising commented:
If I have to merge another sln file this week, I may go postal.

oising committed:
merge upstream main; refactor AEH resource into distinct project

oising committed:
remove azure provisioner credential configuration stuff (will go into separate PR)

oising commented:
> The grand split has happened!

Merged.

oising committed:
merge main

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/Aspire.Azure.Messaging.EventHubs.csproj:
```diff
@@ -0,0 +1,29 @@
+
+  <ItemGroup>
+    <PackageReference Include="Azure.Identity" />
+    <PackageReference Include="Azure.Messaging.EventHubs" />
+    <PackageReference Include="Azure.Messaging.EventHubs.Processor" />
```
Do we need to explicitly reference both? Why not just `Azure.Messaging.EventHubs.Processor` which brings in `Azure.Messaging.EventHubs`. That way we only need to maintain 1 version in the Directory.Packages.props.

eerhardt commented on playground/AspireEventHub/README.md:
```diff
@@ -0,0 +1,7 @@
+# EventHubs + Aspire
+
+A project showing how to use the new `AzureBicepResource` to deploy resources that don't exist in aspire by default. 
```
Needs to be updated.

davidfowl replied:
@oising Did you do this?

oising replied:
Deleted

eerhardt commented on playground/AspireEventHub/README.md:
```diff
@@ -0,0 +1,7 @@
+
+A project showing how to use the new `AzureBicepResource` to deploy resources that don't exist in aspire by default. 
+
+This is using a [preview 4 build of .NET Aspire](https://github.com/dotnet/aspire/blob/main/docs/using-latest-daily.md).
```
```suggestion
```

This uses the code that is in the repo. This line is wrong and is just going to get out of date. Can be removed.

eerhardt commented on src/Aspire.Hosting.Azure.EventHubs/Aspire.Hosting.Azure.EventHubs.csproj:
```diff
@@ -0,0 +1,19 @@
+  <PropertyGroup>
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <PackageTags>aspire hosting azure</PackageTags>
```
```suggestion
    <PackageTags>aspire hosting azure eventhubs</PackageTags>
```

eerhardt commented on src/Aspire.Hosting.Azure.EventHubs/Aspire.Hosting.Azure.EventHubs.csproj:
```diff
@@ -0,0 +1,19 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <InternalsVisibleTo Include="Aspire.Hosting.Tests" />
```
Is this necessary?

eerhardt commented on src/Aspire.Hosting.Azure.EventHubs/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+namespace Aspire.Hosting;
+
+/// <summary>
+/// Provides extension methods for adding the Azure Service Bus resources to the application model.
```
Needs to be updated.

eerhardt commented on playground/AspireEventHub/EventHubs.AppHost/Program.cs:
```diff
@@ -0,0 +1,18 @@
+var builder = DistributedApplication.CreateBuilder(args);
+
+builder.AddAzureProvisioning();
```
Can you generate and check in a manifest and bicep file for the playground? You can do this by:

```shell
cd playground/AspireEventHubs/EventHubs.AppHost
dotnet run -- --publisher manifest --output-path aspire-manifest.json
```

eerhardt commented on src/Aspire.Hosting.Azure.EventHubs/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <param name="builder">The Azure Event Hubs resource builder.</param>
+    /// <param name="name">The name of the Event Hub.</param>
+    /// <param name="configureHub">Optional callback to customize the Event Hub.</param>
+    [RequiresPreviewFeatures]
```
These should be `[Experimental]` instead. See https://github.com/dotnet/aspire/issues/2918

mitchdenny replied:
This should be `[Experimental]`

mitchdenny commented on playground/AspireEventHub/EventHubs.AppHost/appsettings.Development.json:
```diff
@@ -10,5 +10,6 @@
     "AllowResourceGroupCreation": true
-  }
+  },
+  "AzureProvisionerCredential": "AzureCli"
```
Undo

eerhardt commented on src/Aspire.Hosting.Azure.EventHubs/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="configureResource">Optional callback to configure the Event Hubs namespace.</param>
+    /// <returns></returns>
+    [RequiresPreviewFeatures]
```
These should be `[Experimental]` instead. See https://github.com/dotnet/aspire/issues/2918

mitchdenny replied:
This should be `[Experimental]`

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/Aspire.Azure.Messaging.EventHubs.csproj:
```diff
@@ -0,0 +1,29 @@
+  <PropertyGroup>
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+    <IsPackable>true</IsPackable>
+    <PackageTags>$(ComponentAzurePackageTags) servicebus amqp</PackageTags>
```
Needs to be updated.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    // default Tracing to true if the experimental switch is set
+    // TODO: remove this when ActivitySource support is no longer experimental
```
Is tracing for EventHubs still experimental, just like ServiceBus?

cc @lmolkova

lmolkova replied:
yep, it is still experimental

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -0,0 +1,115 @@
+
+    // default Tracing to true if the experimental switch is set
+    // TODO: remove this when ActivitySource support is no longer experimental
+    private static bool GetTracingDefaultValue()
```
Can we share this code with ServiceBus instead of duplicating it?

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -0,0 +1,115 @@
+    public string? ConnectionString { get; set; }
+
+    /// <summary>
+    /// Gets or sets the fully qualified Service Bus namespace. 
```
This needs to be updated.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -0,0 +1,115 @@
+    private bool? _tracing;
+
+    /// <summary>
+    /// Gets or sets the connection string used to connect to the Service Bus namespace. 
```
Needs to be updated.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -0,0 +1,115 @@
+namespace Aspire.Azure.Messaging.EventHubs;
+
+/// <summary>
+/// Provides the client configuration settings for connecting to Azure Service Bus.
```
Needs to be updated.

eerhardt commented on playground/AspireEventHub/EventHubs.AppHost/EventHubs.AppHost.csproj:
```diff
@@ -0,0 +1,39 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="appsettings.Development.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="appsettings.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+
+
```
```suggestion
```

Also please format the whitespace in this file.

eerhardt commented on playground/AspireEventHub/EventHubs.AppHost/EventHubs.AppHost.csproj:
```diff
@@ -0,0 +1,39 @@
+  <ProjectReference Include="..\..\..\src\Aspire.Hosting.Azure.EventHubs\Aspire.Hosting.Azure.EventHubs.csproj" IsAspireProjectResource="false" />
+
+
+  <ProjectReference Include="..\..\..\src\Aspire.Hosting.Azure.Storage\Aspire.Hosting.Azure.Storage.csproj" IsAspireProjectResource="false" />
```
Why do we need storage?

oising replied:
To create a blob container for checkpointing with the EventProcessorClient (see in the consumer project for the processor component settings)

eerhardt commented on playground/AspireEventHub/EventHubsApi/Program.cs:
```diff
@@ -0,0 +1,28 @@
+builder.AddAzureEventHubProducerClient("eventhubns", settings =>
+{
+    settings.EventHubName = "hub";
+    settings.Credential = new AzureCliCredential();
```
```suggestion
```

This should stay `null` and just use `DefaultAzureCredential` when it isn't specified.

eerhardt commented on playground/AspireEventHub/EventHubsConsumer/Program.cs:
```diff
@@ -0,0 +1,44 @@
+}
+else
+{
+    // FIXME: this DI instance is never used as it's not reachable from the processor client configurator; we only use the connection name.
```
What's the plan to address this?

eerhardt commented on playground/AspireEventHub/EventHubsConsumer/Processor.cs:
```diff
@@ -0,0 +1,60 @@
+/// <param name="logger"></param>
+internal sealed class Processor(EventProcessorClient client, ILogger<Consumer> logger) : BackgroundService
+{
+    public override async Task StartAsync(CancellationToken cancellationToken)
```
Typically you don't override both Start and Execute in BackgroundService.

Check out 

https://github.com/dotnet/aspire/blob/166eb52605e1a5190fa63d00142bee57cb79c52f/playground/TestShop/OrderProcessor/OrderProcessingWorker.cs#L9-L29

for an example.

eerhardt commented on playground/AspireEventHub/EventHubsConsumer/EventHubsConsumer.csproj:
```diff
@@ -0,0 +1,21 @@
+    <UserSecretsId>dotnet-WorkerService1-9c946452-f682-49a4-9209-29ed63174447</UserSecretsId>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Azure.Identity" />
+    <PackageReference Include="Microsoft.Extensions.Hosting" />
+  </ItemGroup>
+
```
```suggestion
```

These come transitively through EventHubs reference. Not need for them.

eerhardt commented on playground/AspireEventHub/EventHubsConsumer/EventHubsConsumer.csproj:
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
```suggestion
<Project Sdk="Microsoft.NET.Sdk.Worker">
```

Then you can also delete the `<OutputType>exe</OutputType>` line below as well.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/AspireEventHubsExtensions.cs:
```diff
@@ -0,0 +1,241 @@
+    {
+        // configure processor identifier
+        var slug = Guid.NewGuid().ToString().Substring(24);
+        var identifier = $"{Environment.MachineName}-{settings.EventHubName}-" +
```
Is putting the MachineName in here correct? Especially if these things get deployed to ACA or some other K8s environment.

oising replied:
Also, I like the machine name because it is a good reminder that things like pod anti-affinity exist for improving isolation and performance of a heavily loaded system.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -0,0 +1,93 @@
+                throw new InvalidOperationException(
+                    "A EventProcessorClient could not be configured. " +
+                    $"There is no connection string in Configuration with the name {settings.BlobClientConnectionName}. " +
+                    "Are you sure you defined a Blob resource with this name in your AppHost?");
```
This message is kind of weird. These components can be used outside of an Aspire AppHost. Maybe word to it to say "Ensure you have configured an Azure Storage Blob connection ....".

oising replied:
Good idea.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -0,0 +1,93 @@
+
+        // consumer group and blob container names have similar constraints (alphanumeric, hyphen) but we should sanitize nonetheless
+        var suffix = (string.IsNullOrWhiteSpace(settings.ConsumerGroup)) ? "default" : settings.ConsumerGroup;
+        blobUriBuilder.BlobContainerName = $"checkpoints-{suffix}"; // TODO: should be unique to this consumer group
```
What's the plan to address this TODO?

oising replied:
Address elsewhere with @jsquire

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -0,0 +1,93 @@
+                    $"There is no connection string in Configuration with the name {settings.BlobClientConnectionName}. " +
+                    "Are you sure you defined a Blob resource with this name in your AppHost?");
+
+        // FIXME: ideally this should be pulled from services; but thar be dragons.
+        // There is no reliable way to get the blob service client from the services collection
```
Why not? Why not grab the `BlobContainerClient` or `BlobServiceClient` from DI, and if it isn't there, throw? That's the whole point of DI... to inject dependencies.

oising replied:
There is no service provider in scope here. Also, if it was added without a ServiceKey, how do I know if it's the one I need? This needs more thought.

oising replied:
I am going to remove the blob component code from the worker in this sample as it is actually doing nothing for these very reasons except seed confusion. I agree @eerhardt that we _should_ be using the DI instance if it's there, but the plumbing isn't there yet.

oising replied:
Also, pulling from DI (assuming a service provider is available) puts an ordering dependency on the components that may not be obvious without clear documentation.

oising replied:
removed blob component from sample eventhub (but it remains as a resource in apphost)

eerhardt replied:
Can you open an issue for this? I think this is going to need to be reworked before this component can be GA'd. Developers are going to want to configure this Blob client more than just passing the "connection name" (which according to the current code can only be a URI).

These components need to be "composable". That's the whole point.

mitchdenny commented:
The AppModel side of this looks OK other than the preview attributes needing to be changed.

jsquire commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -0,0 +1,93 @@
+
+        // consumer group and blob container names have similar constraints (alphanumeric, hyphen) but we should sanitize nonetheless
+        var suffix = (string.IsNullOrWhiteSpace(settings.ConsumerGroup)) ? "default" : settings.ConsumerGroup;
+        blobUriBuilder.BlobContainerName = $"checkpoints-{suffix}"; // TODO: should be unique to this consumer group
```
I'd suggest changing the container name to something like `"{namespace}{eventhub}{consumergroup}`, as it will contain both checkpoints and ownership information for processors in the cluster.  The processor will create Blob Storage pseudo-folder thingies for "checkpoints" and "ownership" in the container.  The limitations on Blob container names make it hard to do this both readable and unique.

I don't think you should assume that you want to create a brand new unique container for every component.  You want the processors to be able to easily join a cluster for the same namespace/Event Hub/consumer group combination, as that is necessary for effective scaling.  The processor has an unbounded per-partition concurrency model and for most real-world scenarios, an application will need to have multiple processors sharing work to avoid overloading a single host.

timheuer commented on src/Aspire.Hosting.Azure.EventHubs/Aspire.Hosting.Azure.EventHubs.csproj:
Please add image (attaching official icon here to put in shared) and README.md to the hosting package.  
![AzureEventHubs_256x](https://github.com/dotnet/aspire/assets/4821/67cf1f13-8cd0-4bfa-90a4-178edfa7b8c5)

oising committed:
address review points; refactor namespace parsing into base eh component

oising committed:
fix package tags and add new icon to shared

oising committed:
remove providercredential reference from config

oising committed:
merge main

oising committed:
fix some more errant service bus mentions and regenerate configurationschema

oising committed:
fix xmlcomment on settings

oising committed:
remove preference for AzureCliCredential in sample; remove my subscriptionId lol

oising committed:
remove azure section from sample config

oising committed:
fix build

davidfowl commented on playground/AspireEventHub/README.md:
```diff
@@ -0,0 +1,7 @@
+
+This is using a [preview 4 build of .NET Aspire](https://github.com/dotnet/aspire/blob/main/docs/using-latest-daily.md).
+
+This project requires an [a daily build](https://github.com/Azure/azure-dev/blob/main/cli/installer/README.md#download-from-daily-builds) of azd to deploy to azure.
```
This entire readme can be removed.

oising committed:
update components README.md

oising committed:
delete EH playground readme

oising commented on src/Components/Aspire.Azure.Messaging.EventHubs/README.md:
````diff
@@ -0,0 +1,145 @@
+In your AppHost project, install the Aspire Azure Hosting library with [NuGet](https://www.nuget.org):
+
+```dotnetcli
+dotnet add package Aspire.Hosting.Azure
````
@davidfowl @eerhardt Checked in and ready for first review.

oising committed:
minor edits for clarification

oising committed:
port ASB connection/namespace tests; rewrite validation logic to be more robust

oising committed:
Update playground/AspireEventHub/EventHubs.AppHost/EventHubs.AppHost.csproj

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

oising committed:
Update playground/AspireEventHub/EventHubsConsumer/EventHubsConsumer.csproj

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

oising committed:
Update src/Aspire.Hosting.Azure.EventHubs/Aspire.Hosting.Azure.EventHubs.csproj

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

oising committed:
generate aspire manifest etc for EH sample

oising committed:
Merge branch 'support-eventhubs' of https://github.com/oising/aspire into support-eventhubs

oising committed:
moved processor Start code into Execute for worker sample

oising committed:
fixed another ref to service bus

oising committed:
refine checkpoint blob container creation logic to avoid unneccessary permission demand if we can

oising committed:
enhance logic for blob checkpoint container; add BlobContainerName to settings as an option

eerhardt commented:
It looks like the tests are failing. @oising - can you investigate?

oising committed:
fix borked test

oising committed:
alright markdown linter, you win this time

eerhardt commented on playground/AspireEventHub/EventHubs.AppHost/Program.cs:
```diff
@@ -0,0 +1,19 @@
+var blob = builder.AddAzureStorage("ehstorage")
+    .AddBlobs("checkpoints");
+
+var eventHub = builder.AddAzureEventHubs("eventhubns")
+    .AddEventHub("hub");
```
@jsquire - should this be:

```C#
var eventHub = builder.AddAzureEventHubNamespace("eventhubns")
    .AddEventHub("hub");
```

Looking at the portal and in the bicep, I think this top level thing is called a "Namespace". Should we be reflecting that in the API?

davidfowl replied:
I think we should file a follow up to do this for both service bus and this resource.

jsquire replied:
Technically, yes.   You're correct about the nomenclature.  I do think it adds a bit of clarity, though I didn't find it difficult to understand in the current form, either.

oising replied:
I think that would be good, but as others have said, ASB should reflect the same nomenclature. As it stands, I'm doing what the service bus stuff does.

jsquire reviewed (approved):
Event Hubs parts look good to me.

jsquire commented on playground/AspireEventHub/EventHubsConsumer/Consumer.cs:
```diff
@@ -0,0 +1,33 @@
+/// <summary>
+///   Demonstrate consuming events from an Event Hub using the <see cref="EventHubConsumerClient" />.
+/// </summary>
+/// <param name="client"></param>
```
Do we want to fill these out?

eerhardt replied:
The code under "playground" is just manual test apps. XML doc comments in that folder aren't required/used.

oising commented on tests/Aspire.Azure.Messaging.EventHubs.Tests/AspireEventHubsExtensionsTests.cs:
```diff
@@ -0,0 +1,292 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
I'm back to a full work week this week, so I've only got a couple of evenings free before code freeze for P5, so I'll need some guidance here to be efficient. I thought the conformance stuff is based on pre-existing internal MSFT azure infra, things that I can neither see nor access. Is this even something I can do?

eerhardt replied:
>  so I'll need some guidance here to be efficient

Copy the ConformanceTests from another component (storage or service bus), and then update it to EventHubs.

> I thought the conformance stuff is based on pre-existing internal MSFT azure infra, things that I can neither see nor access. Is this even something I can do?

There are a bunch of "Conformance Tests" that can be done without even connecting to Azure. These are the tests that run in CI, which also don't have the ability to connect to Azure.

For the tests that do need to connect to Azure, you can develop and run them locally, pointing to a real Azure EventHubs instance that you have access to. And then when you push the change to the PR, change the EventHub URL. See 

https://github.com/dotnet/aspire/blob/aeeee41cc9105eb82ab513b2a9a239a67666b8a2/tests/Aspire.Azure.Messaging.ServiceBus.Tests/ConformanceTests.cs#L16

oising commented on tests/Aspire.Azure.Messaging.EventHubs.Tests/AzureMessagingServiceBusSettingsTests.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Aspire.Azure.Messaging.EventHubs.Tests;
+
+public class AzureMessagingServiceBusSettingsTests
```
Oops.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/README.md:
```diff
@@ -0,0 +1,182 @@
+The shared settings for all clients are suffixed below with one or more
+indices corresponding to the client types listed above that may use them:
+
+| Member                                                                                   | Type             |
+|------------------------------------------------------------------------------------------|------------------|
+| ConnectionString[1,2,3,4]   | string?          |
+| Namespace[1,2,3,4]          | string?          |
+| EventHubName[1,2,3,4]       | string?          |
+| ConsumerGroup[2,3,4]        | string?          |
+| BlobClientConnectionName[3] | string?          |
+| BlobContainerName[3]        | string?          |
+| PartitionId[4]              | string?          |
+| EventPosition[4]            | EventPosition    |
+| Credential[1,2,3,4]         | TokenCredential? |
+| Tracing[1,2,3,4]            | bool             |
```
Would it make more sense to split these into separate classes? That way we didn't have properties on Settings types that didn't work?

For example, I could imagine this hierarchy:

```C#
abstract class AzureMessagingEventHubsBaseSettings
{
  ConnectionString
  Namespace
  EventHubName
  Credential
  Tracing
}

sealed class AzureMessagingEventHubsProducerSettings : AzureMessagingEventHubsBaseSettings {}

abstract class AzureMessagingEventHubsConsumerBaseSettings: AzureMessagingEventHubsBaseSettings
{
  ConsumerGroup
}

sealed class AzureMessagingEventHubsConsumerSettings : AzureMessagingEventHubsBaseSettings {}

sealed class AzureMessagingEventHubsProcessorSettings : AzureMessagingEventHubsBaseSettings
{
  BlobClientConnectionName
  BlobContainerName
}

sealed class AzureMessagingEventHubsPartitionReceiverSettings : AzureMessagingEventHubsBaseSettings
{
  PartitionId
  EVentPosition
}
```

oising replied:
Absolutely makes sense from an objective, clean separation of concerns point of view. But from an Aspire point of view where it would mean delving into the Aspire.Hosting common framework to modify it to handle multiple types of settings classes, I don't know if this would be in scope for this PR. A reminder that this is the only componenty (Event Hubs) that has multiple client types; all others have a single client type, i.e. for ASB, it only has ServiceBusSender. This is the only way I could see to make this work within the p5 timeline. Feel free to bring this up internally with @davidfowl and @mitchdenny

eerhardt replied:
> from an Aspire point of view where it would seem to mean (afaict) delving into the Aspire.Hosting common framework to modify it to handle multiple types of settings classes

I don't follow this comment. "Aspire.Hosting" doesn't matter here. Each public API in the component can take its own Settings type:

``` 
public static void AddAzureEventProcessorClient(
     this IHostApplicationBuilder builder,
     string connectionName,
     Action<AzureMessagingEventHubsProcessorSettings>? configureSettings = null

...

    public static void AddAzureEventHubProducerClient(
        this IHostApplicationBuilder builder,
        string connectionName,
        Action<AzureMessagingEventHubsProducerSettings>? configureSettings = null
```

oising replied:
Ah, I see -- thanks for the tip. Apologies -- I've only had my head in this code for the best part of a long weekend.

oising replied:
@eerhardt Is there a way to flag the base settings types so they don't get picked up by the ConfigurationSchema source generator? Or does this even matter?

oising replied:
Never mind -- I assume it will ignore abstract types.

eerhardt replied:
In general the ConfigurationSchema generator will skip the properties/types that don't work with the config binder source generator. There is a way to make the schema generator skip properties if you need to manually say to skip them:

https://github.com/dotnet/aspire/blob/23104ee048dcf99bc1bdcb5ff8ad01ee85921d3b/src/Components/Common/ConfigurationSchemaAttributes.cs#L29-L32

oising replied:
Done.

eerhardt commented on src/Components/Aspire.Azure.Messaging.EventHubs/README.md:
```diff
@@ -0,0 +1,182 @@
+The following example assumes that you have an Azure Event Hubs namespace and an Event Hub created and wish to configure an `EventHubProducerClient` to send events to the Event Hub. The `EventHubConsumerClient`, `EventProcessorClient`, and `PartitionReceiver`are configured in a similar manner.
+
+In the _Program.cs_ file of your project, call the `AddAzureEventHubProducerClient` extension method to register
+a `EventHubProducerClient` for use via the dependency injection container. The method takes a connection name parameter. This assumes you have included the `EntityPath` in the connection string to specify the Event Hub name.
```
> This assumes you have included the `EntityPath` in the connection string to specify the Event Hub name.

This doesn't seem ideal, since we say we don't recommend using a connection string.

eerhardt replied:
It is also surprising since I would have expected:

```
var eventHub = builder.AddAzureEventHubs("eventhubns")
    .AddEventHub("hub");

builder.AddProject<Projects.EventHubsApi>("api")
    .WithReference(eventHub);
```

To result with the `EntityPath` in the connection string (just like it does if I'm using a database server + database, the connection string has `Database=<db_name>` in the connection string.

Is the above code supposed to use `EntityPath` in the connection string?

oising replied:
@mitchdenny Brought this up already and it would mean making the "hub" it's own de-facto AspireResource to enable deep-linking like this. I replied and said the same could be said for allowing deep-linking of ASB topic and queues (a feature that isn't there either) so I think this could be punted for another release. 

As you (or maybe someone else said?) before, we do control what Resources inject as connection strings when using an AppHost, but we don't control what developers may do when managing their own connection strings. I'd rather that we "do the right thing" when a dev passes an entitypath. It seems like the path of least surprise. It's what I would expect.

eerhardt replied:
> we don't control what developers may do when managing their own connection strings. I'd rather that we "do the right thing" when a dev passes an entitypath. It seems like the path of least surprise. It's what I would expect.

I don't disagree that we should "do the right thing" when someone passes an entitypath.

The part I was commenting on is the "Usage example" the README gives assumes you are doing something we don't recommend you do - use a connection string.

oising replied:
Ah, gotcha!

eerhardt commented on playground/AspireEventHub/EventHubsConsumer/Processor.cs:
```diff
@@ -0,0 +1,55 @@
+    {
+        logger.LogInformation("Starting processor...");
+
+        client.ProcessEventAsync += arg =>
+        {
+            logger.LogInformation(arg.Data.EventBody.ToString());
+            return Task.CompletedTask;
+        };
```
When playing with this app, it seems that every time it starts up, it replays all the events that were already produced from previous runs. Is that intentional?

oising replied:
Seems like a bug -- it should be checkpointing. Good catch; I may have broken something recently.

oising replied:
Oh, it was never checkpointing -- there's no call to UpdateCheckpointAsync. Normally this is called after a success processing of the batch. I'm going to add this line with an explanatory comment. Good catch nonetheless.

eerhardt commented on src/Aspire.Hosting.Azure.EventHubs/AzureEventHubsExtensions.cs:
```diff
@@ -0,0 +1,101 @@
+    /// </summary>
+    /// <param name="builder">The Azure Event Hubs resource builder.</param>
+    /// <param name="name">The name of the Event Hub.</param>
+    public static IResourceBuilder<AzureEventHubsResource> AddEventHub(this IResourceBuilder<AzureEventHubsResource> builder, string name)
```
This API pattern differs from how databases work. I think it will be surprising to our customers.

In databases, we have:

```C#
var db = builder.AddPostgres("db-server").AddDatabase("db1");

builder.AddProject<Projects.Api>("api")
    .WithReference(db);
```

When executing that code, `db` points to the `db1` resource - a "child" of the `db-server`. So when you `.WithReference(db)`, the connection string name that you use in your app is going to be `db1`.

But here,

```C#
var hub = builder.AddEventHubs("eventhub-ns").AddEventHub("hub");

builder.AddProject<Projects.Api>("api")
    .WithReference(hub);
```

`hub` is going to point to the `eventhub-ns` resource, and not the `hub` resource. So in your web app, you need to use `builder.AddAzureEventHubProducerClient("eventhub-ns")`, which is different than how databases work.

oising replied:
Agree 100% -- see my comment above. The service bus resource is equally deficient, and I copied that code. I'd want to see this fixes for both ASB and AEH in P6.  /cc @mitchdenny

davidfowl replied:
Does this work with managed identity?

oising replied:
> Does this work with managed identity?

Did you mean to put this here, @davidfowl ? If so, I don't see why not?

oising committed:
persist checkpoints in processor sample for EH

oising committed:
renamed settings class in tests to match AEH; was ASB.

oising committed:
doh. forgot to save the file

eerhardt committed:
Fix Tracing with EventHubs
Update Telemetry and Components Progress docs
Minor cleanup feedback

eerhardt commented:
FYI @oising, I pushed some changes you should be aware of. The most important is that I fixed Tracing to work. Just wanted to let you know so you don't force push and revert those changes.

I think this is close to getting the initial merge. My most major concerns are:

* Getting the Storage Blob client. I really think this should come from DI
* The Settings class having properties that don't work depending on the client being used.
* Full testing

If we want to merge this soon to guarantee it gets into preview5, and log issues for these to be followed up, I would be OK with that. FYI @davidfowl

oising commented:
> FYI @oising, I pushed some changes you should be aware of. The most important is that I fixed Tracing to work. Just wanted to let you know so you don't force push and revert those changes.
> 
> I think this is close to getting the initial merge. My most major concerns are:
> 
> * Getting the Storage Blob client. I really think this should come from DI
> * The Settings class having properties that don't work depending on the client being used.
> * Full testing
> 
> If we want to merge this soon to guarantee it gets into preview5, and log issues for these to be followed up, I would be OK with that. FYI @davidfowl

Thanks @eerhardt -- I really appreciate the thorough review and inciteful suggestions :) I'm about to check in the suggested split settings approach as discussed above. It took a bit of jiggery-pokery with generics to fix a few things, but it seems to be good now.

As for the storage blob client, yes, I agree this is a bit of a rough edge; especially around the lack of unit tests here. As mentioned, the service provider isn't built at this point, but I could build it and look for a blob client within. I think this needs some configuration-by-convention love in the sense that we should probably only permit a keyed blob client so it can be configured in settings. Also, to test the logic around the container creation, I might want to lightly shim the blob client so I can unit test the behaviour without requiring a live connection. I'll spike this tomorrow.

oising committed:
refactor EH client settings into individual classes

oising committed:
merge main

oising committed:
Merge branch 'main' into support-eventhubs

oising commented on src/Aspire.Hosting.Azure.EventHubs/Aspire.Hosting.Azure.EventHubs.csproj:
```diff
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
On this today.

oising replied:
Done

oising committed:
Add readme for EH hosting

eerhardt commented:
> As mentioned, the service provider isn't built at this point, but I could build it and look for a blob client within.

This isn't necessary. We should just be able to call a different overload on the AzureClientFactoryBuilder, which gives us the `IServiceProvider`. See https://github.com/Azure/azure-sdk-for-net/blob/87d426597b261b88f8a6c6c21fab6031f857fb70/sdk/extensions/Microsoft.Extensions.Azure/src/AzureClientFactoryBuilder.cs#L164-L172. (Note it would take some refactoring of the base `AzureComponent<TSettings, TClient, TClientOptions>` class, as right now the `abstract AddClient` method takes a `TBuilder`, when it could just be passed an `AzureClientFactoryBuilder`.

> I think this needs some configuration-by-convention love in the sense that we should probably only permit a keyed blob client so it can be configured in settings.

I see it a different way. It should allow for a keyed blob client, but shouldn't require it to be keyed. If no "blob service key" was configured, it should just get the `BlobServiceClient` from DI. If a "blob service key" was specified, then use it to get the `BlobServiceClient` from DI.

That way the simple case "just works". You just need to:

```C#
var builder = WebApplication.CreateBuilder(args);

builder.AddAzureBlobClient("blobConnectionName");
builder.AddAzureEventProcessorClient("eventhubns");
```

and the EventProcessorClient grabs the BlobClient. There's no need to force someone to use keys when there is only one.

eerhardt committed:
Fix build

eerhardt committed:
Address PR feedback

- Remove unnecessary Directory.Packages.props entry
- Fix ConfigurationSchema.json to match implementation
- StringComparison.Ordinal is unnecessary when looking for a char

eerhardt reviewed (approved):
Thank you, @oising, for this contribution! Great work!

This LGTM to merge for preview5.

Can you log follow up issues for the remaining work? Notably:

* Using DI to get the blob client in the EventProcessor
* Testing
* (I might have missed others. Please ensure it is all logged.)

davidfowl commented:
Thank you @oising !!!!!
